controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -0,0 +1,349 @@
+                            .thenCompose(txnList -> {
+                                if (txnList == null) {
+                                    // reset state conditionally in case we were left with stale committing state from a previous execution
+                                    // that died just before updating the state back to ACTIVE but after having completed all the work.
+                                    return streamMetadataStore.resetStateConditionally(scope, stream, State.COMMITTING_TXN, context, executor);
+                                } else {
+                                    // Once state is set to committing, we are guaranteed that this will be the only processing that can happen on the stream
+                                    // and we can proceed with committing outstanding transactions collected in the txnList step.
+                                    CompletableFuture<Void> future;
+                                    // if state is sealing, we should continue with commit so that we allow for completion of transactions
+                                    // in commit state.
+                                    if (state.equals(State.SEALING)) {
+                                        future = new CompletableFuture<>();
+                                    } else {
+                                        // In normal course set the state to committing before proceeding.
+                                        // If we are unable to set the state to COMMITTING_TXN, it will get OPERATION_NOT_ALLOWED
+                                        // and the processing will be retried later.
+                                        future = Futures.toVoid(streamMetadataStore.setState(scope, stream, State.COMMITTING_TXN, context, executor));
+                                    }
+
+                                    // Note: since we have set the state to COMMITTING_TXN (or it was already sealing), the active epoch that we fetch now
+                                    // cannot change until we perform rolling txn. TxnCommittingRecord ensures no other rollingTxn
+                                    // can run concurrently
+                                    return future.thenCompose(v -> getEpochRecords(scope, stream, txnEpoch, context)
+                                            .thenCompose(records -> {
+                                                HistoryRecord txnEpochRecord = records.get(0);
+                                                HistoryRecord activeEpochRecord = records.get(1);
+                                                if (activeEpochRecord.getEpoch() == txnEpoch ||
+                                                        activeEpochRecord.getReferenceEpoch() == txnEpochRecord.getReferenceEpoch()) {
+                                                    // if transactions were created on or a duplicate of active epoch,
```

Comment:
I think the phrasing here is incorrect. It is not a duplicate of the active epoch. Instead, It think it is the case that the active epoch is a duplicate of the txn reference epoch.